File: problems/valid_anagram/solution_2/python/main.py
def isAnagram(s: str, t: str) -> bool:
    if len(s) != len(t):
        return False

    count = {}
    
    for char in s:
        count[char] = count.get(char, 0) + 1

    for char in t:
        if char not in count:
            return False

        count[char] -= 1

        if count[char] < 0:
            return False

    return True

def testAnagram():
    testCases = [
        ("anagram", "nagaram", True),
        ("rat", "car", False),
    ]

    for s, t, expected in testCases:
        result = isAnagram(s, t)
        print(f"isAnagram({s}, {t}) = {result}, expected = {expected}")

    
testAnagram()

File: problems/valid_anagram/solution_2/golang/main.go
package main

import "fmt"

func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	count := make(map[rune]int)

	for _, char := range s {
		count[char]++
	}

	for _, char := range t {
		count[char]--
		if count[char] < 0 {
			return false
		}
	}

	return true
}

func testAnagram() {
	testCases := []struct {
		s        string
		t        string
		expected bool
	}{
		{"anagram", "nagaram", true},
		{"rat", "car", false},
	}

	for _, tc := range testCases {
		result := isAnagram(tc.s, tc.t)
		fmt.Printf(
			"isAnagram(%q, %q) = %v | Expected: %v\n",
			tc.s,
			tc.t,
			result,
			tc.expected,
		)
	}
}

func main() {
	testAnagram()
}